### DISCUSSION TITLE
How to create a multi-language website with SvelteKit 1.0 and typesafe-i18n

### DISCUSSION THREAD

**Discussion by ciscoheat** [ivanhofer/typesafe-i18n#564] [Show and tell]
This article helps you get over the "setup hurdle", using the latest SvelteKit to make i18n quite simple: 
https://blog.encodeart.dev/typesafe-i18n-with-sveltekit

**Comment by ivanhofer** [comment]
Hi @ciscoheat thanks for sharing this!
I really like your article 👏

May I add just two small notes:
 - it would be great if you can add a link to the repo where the example from the article gets fully shown. Some people might not follow along when they see just small code snippets. The full example would help them.
 - I would rewrite [the matcher](https://blog.encodeart.dev/typesafe-i18n-with-sveltekit#heading-matching-only-the-defined-languages) like this:
	 ```diff
	 import type { ParamMatcher } from '@sveltejs/kit';
    +import { locales } from '$lib/i18n/i18n-util'
	
	export const match: ParamMatcher = (param) => {
	  // No need to match 'en' since it's the default.
	-  return param === 'de';
	+  return locales.includes(param);
	};
	 ``` 
	 this ensures that the matcher will still work when you add a new locale

**Reply by ciscoheat** [reply]
Glad you like it! :) I'll upload the code to github and link it the article soon.

It's funny that you mention the matcher, Initially I wrote it just like you suggest, but yesterday saw a video about how slow Map operations are compared to Array lookups for small datasets, so I took it even further to avoid the array lookup. A quite ridiculous optimization, I know!

**Reply by ivanhofer** [reply]
Interesting..
I just tested a small example on https://jsbench.me/ but could not find any meningful performance differences when using 10 locales.

![image](https://user-images.githubusercontent.com/21335119/211348210-52765672-9e25-43a1-8916-cf36c04eb353.png)

Only if you have a lot of locales, the array lookup will become slightly slower.

**Reply by ciscoheat** [reply]
Exactly, it makes no sense in this case. I will change back to the one you suggested.

**Reply by ciscoheat** [reply]
Btw, since `locales` is strongly typed, is there any way of making includes work without casting either the array or the (string) parameter? I'm doing this currently:

```typescript
return (locales as string[]).includes(param);
```

**Reply by ivanhofer** [reply]
Unfortunately not. But I would cast the other part
```ts
return locales.includes(param as Locales);
```

**Reply by ciscoheat** [reply]
I just realized that I can't use `locales.includes` directly, since `en` is not a valid route. So it has to be:

```typescript
return param !== 'en' && locales.includes(param as Locales);
```

I have cleaned up the code a bit and will put a link to the repo later today.

**Reply by ivanhofer** [reply]
instead of the hardcoded string `'en'`, you could use the `baseLocale` variable :)

**Reply by ciscoheat** [reply]
The imports are getting larger than the code itself, but ok. :)

**Reply by ivanhofer** [reply]
I would see this as a good sign. Less code => less errors ^^
